Geometry_Engine/Modify/SplitAtPoints.cs

```diff
@@ -50,41 +50,146 @@ public static List<Arc> SplitAtPoints(this Arc arc, List<Point> points, double t
 
             List<Arc> result = new List<Arc>();
             List<Point> cPts = new List<Point>();
-            Vector normal = arc.Normal();
 
+            double sqTol = tolerance * tolerance;
+            Plane arcPlane = arc.FitPlane();
+            //Collect all points on the circle of the arc (points not on arc filtered out lower down)
+            //Doing this half manually (Not calling IsOnCurve) due to wanting the projected points anyway, and this is more efficient.
+            //Angles checked to be in domain while looping through the points further down the algorithm
             foreach (Point point in points)
             {
-                if (point.IsOnCurve(arc, tolerance))
-                    cPts.Add(point);
-            }
+                Point projPt = point.Project(arcPlane);
+                double projSqDist = projPt.SquareDistance(point);
+                if (projSqDist > sqTol)
+                    continue;   //Not in plane
 
-            cPts.Add(arc.StartPoint());
-            cPts.Add(arc.EndPoint());
-            cPts = cPts.CullDuplicates(tolerance);
-            cPts = cPts.SortAlongCurve(arc, tolerance);
+                double inPlaneDist = Math.Abs(projPt.Distance(arcPlane.Origin) - arc.Radius);
+                if (inPlaneDist > tolerance)
+                    continue;   //Not within tolerance distance in plane
 
-            if (arc.EndAngle - 2 * Math.PI < tolerance && arc.EndAngle - 2 * Math.PI > -tolerance)
-                cPts.Add(arc.EndPoint());
+                if (projSqDist + inPlaneDist * inPlaneDist > sqTol) //Distance to the circle of the full arc (c^2=a^2+b^2 where a is out of plane distance and b is in plane distance) 
+                    continue;   //Not on circle
 
-            if (cPts.Count > 2)
+                cPts.Add(projPt);   //Store projected point rather than original for more accurate angle calculation
+            }
+
+            //If any points on the full circle corresponding to the arc
+            if (cPts.Count > 0)
             {
-                Double startAng = arc.StartAngle;
-                Double endAng = arc.EndAngle;
-                Double tmpAng = 0;
-                Arc tmpArc;
+                double angleTolerance = tolerance / arc.Radius; //Allowable difference in angle for the points on the arc to be tolerance distance away from each other
```

Comment:
Are you 100% sure this value corresponds to the distance between two points on the arc (marked in blue)? I am not saying it is not, actually I have a blackout and can't find a definite answer 😆 

![image](https://user-images.githubusercontent.com/26874773/197186138-7ac10dbf-16c7-4d2b-8750-9b0d2b316cdf.png)

Reply:
No, you are right, it is the length of the black, not the blue. Will try to update.